src/aws-cpp-sdk-core/source/smithy/tracing/impl/opentelemetry/MeterAdapters.cpp

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <smithy/tracing/impl/opentelemetry/MeterAdapters.h>
+#include <opentelemetry/metrics/sync_instruments.h>
+#include <opentelemetry/metrics/async_instruments.h>
+#include <opentelemetry/metrics/observer_result.h>
+#include <opentelemetry/context/runtime_context.h>
+#include <opentelemetry/nostd/variant.h>
+
+#include <utility>
+
+using namespace smithy::components::tracing;
+
+static const char *ALLOC_TAG = "OTEL_METER_ADAPTER";
+
+Aws::UniquePtr<MonotonicCounter> OtelMeterAdapter::CreateCounter(Aws::String name,
+    Aws::String units,
+    Aws::String description) {
+    auto counter = otelMeter->CreateUInt64Counter(name, description, units);
+    return Aws::MakeUnique<OtelCounterAdapter>(ALLOC_TAG, std::move(counter));
+}
+
+Aws::UniquePtr<GaugeHandle> OtelMeterAdapter::CreateGauge(Aws::String name,
+    std::function<void(Aws::UniquePtr<AsyncMeasurement>)> callback,
+    Aws::String units,
+    Aws::String description) {
+    auto gauge = otelMeter->CreateInt64ObservableGauge(name, description, units);
+    GaugeHandleState gaugeHandleState{callback};
+    auto callbackFunc = [](opentelemetry::metrics::ObserverResult result, void *state) -> void {
+        Aws::UniquePtr<AsyncMeasurement> measurement = Aws::MakeUnique<OtelObserverAdapter>(ALLOC_TAG, result);
+        auto handleState = (GaugeHandleState *) state;
```

Comment:
I think you should be able to use more specific or C++-style reinterpret_cast (not that this changes much though).
Also, please check that the pointer is not null.